akka-remote/src/main/scala/akka/remote/artery/Association.scala

```diff
@@ -43,15 +43,30 @@ import akka.util.OptionVal
 import org.agrona.concurrent.ManyToOneConcurrentArrayQueue
 import akka.remote.artery.compress.CompressionProtocol._
 import akka.stream.scaladsl.MergeHub
+import java.util.concurrent.atomic.AtomicBoolean
 
 /**
  * INTERNAL API
  */
 private[remote] object Association {
-  final case class QueueWrapper(queue: Queue[OutboundEnvelope]) extends SendQueue.ProducerApi[OutboundEnvelope] {
+  sealed trait QueueWrapper extends SendQueue.ProducerApi[OutboundEnvelope] {
+    def queue: Queue[OutboundEnvelope]
+  }
+
+  final case class QueueWrapperImpl(queue: Queue[OutboundEnvelope]) extends QueueWrapper {
     override def offer(message: OutboundEnvelope): Boolean = queue.offer(message)
   }
 
+  final case class LazyQueueWrapper(queue: Queue[OutboundEnvelope], materialize: () ⇒ Unit) extends QueueWrapper {
+    private val onlyOnce = new AtomicBoolean
+
+    override def offer(message: OutboundEnvelope): Boolean = {
+      if (onlyOnce.compareAndSet(false, true))
+        materialize()
+      queue.offer(message)
```

Comment:
So... technically one materialize() may still be executing while we hit the offer here from a different thread, but that's fine as the queue is shared and the _one_ materialized stream will start reading from it immediately?